### DISCUSSION TITLE
hardware/2024/09/01/qualcomm_oryon/

### DISCUSSION THREAD

**Discussion by giscus** [jiegec/blog-source#15] [General]
# hardware/2024/09/01/qualcomm_oryon/

杰哥的{运维，编程，调板子}小笔记

https://jia.je/hardware/2024/09/01/qualcomm_oryon/

**Comment by xiaoerlang0359** [comment]
请问取指带宽测试中，横坐标表示的是一次循环中的指令数，纵坐标表示一次循环的cycle数是吗？另外，cross-page的执行cycle在15-20中间突然下降是什么原因呢？

**Reply by jiegec** [reply]
> 请问取指带宽测试中，横坐标表示的是一次循环中的指令数，纵坐标表示一次循环的cycle数是吗？

横坐标是第二个页内的指令数，一次循环的指令数等于横坐标+1。纵坐标是运行很多次循环的总 cycle 数除以循环次数。

> 另外，cross-page的执行cycle在15-20中间突然下降是什么原因呢？

这里涉及到两个可能成为瓶颈的地方，一是 Fetch，Oryon 每周期可以取 16 条指令，可以跨缓存行，但不能跨页；二是 Commit，每周期可以 Commit 8 条指令。

cross-page 测试下，从横坐标 <= 15 的区间内是 Fetch 成为瓶颈，跨页的情况下，至少 2 周期，此时执行并不需要 2 周期。横坐标 = 16 时，Fetch 需要 2 周期，Commit 需要 17/8 周期，此时瓶颈在 Commit。横坐标在 [17, 20] 之间时，Fetch 需要 3 周期，Commit 需要 (x+1)/8 周期，此时瓶颈在 Fetch，按理说这一段都应该观察到 3 周期/循环。但比较奇怪的是，x=17 时实测 4 周期/循环，背后的原因目前还没有深入研究。

**Comment by qianwindz** [comment]
命中 L2 TLB 的时间有长有短，说明它的 entry 不是等同的。
这个有没有可能是因为这个l2 tlb 是Unified，l1i tlb refill和l1d tlb refill 都会触发对l2 tlb的访问

**Reply by jiegec** [reply]
l2 tlb 确实是 unified，不过我在这里测的时候，是用测 dtlb 的方法去测 l2 tlb 的容量，也就是说此时指令 footprint 是很小的，按理说只有 l1d tlb refill 而没有 l1i tlb refill。
